### PR TITLE
resolve: mark it undetermined if single import is not has any bindings

### LINKED ISSUE #124490: ICE: `inconsistent resolution for an import`
auto-reduced (treereduce-rust):
````rust
use io::{self as std};
use std::collections::{self as io};


````

original:
````rust
use std::collections::{self as io};
use io::{self as std};

````

Version information
````
rustc 1.80.0-nightly (91d5e4af8 2024-04-28)
binary: rustc
commit-hash: 91d5e4af8674bde06096bac1e7ac04af0e94c691
commit-date: 2024-04-28
host: x86_64-unknown-linux-gnu
release: 1.80.0-nightly
LLVM version: 18.1.4

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Zunstable-options --edition=2024`

<details><summary><strong>Program output</strong></summary>
<p>

```
error: internal compiler error: compiler/rustc_resolve/src/imports.rs:884:25: inconsistent resolution for an import
 --> /tmp/icemaker_global_tempdir.7N3g3X1DaYFn/rustc_testrunner_tmpdir_reporting.hyRGLJm0FCJ6/mvce.rs:2:24
  |
2 | use std::collections::{self as io};
  |                        ^^^^^^^^^^

thread 'rustc' panicked at compiler/rustc_resolve/src/imports.rs:884:25:
Box<dyn Any>
stack backtrace:
   0:     0x7b6b0a636035 - std::backtrace_rs::backtrace::libunwind::trace::ha90ea7a57c64b8ab
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/std/src/../../backtrace/src/backtrace/libunwind.rs:105:5
   1:     0x7b6b0a636035 - std::backtrace_rs::backtrace::trace_unsynchronized::ha58e55f622939b6e
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7b6b0a636035 - std::sys_common::backtrace::_print_fmt::h687007ed504e7e64
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x7b6b0a636035 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hecfae54cdc6fec81
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7b6b0a68529b - core::fmt::rt::Argument::fmt::h1cf6305de42401ed
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/core/src/fmt/rt.rs:165:63
   5:     0x7b6b0a68529b - core::fmt::write::hb74fc275648248fb
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/core/src/fmt/mod.rs:1157:21
   6:     0x7b6b0a62abdf - std::io::Write::write_fmt::h5209ffd8e6502f0c
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/std/src/io/mod.rs:1832:15
   7:     0x7b6b0a635e0e - std::sys_common::backtrace::_print::h168707047d27c84e
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7b6b0a635e0e - std::sys_common::backtrace::print::h2942d174d6b0e46d
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7b6b0a638779 - std::panicking::default_hook::{{closure}}::h4215c46ca923b014
  10:     0x7b6b0a6384bd - std::panicking::default_hook::h887f937d01fa1b64
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/std/src/panicking.rs:298:9
  11:     0x7b6b06e40a1c - std[aaff33ccdb7ea7e2]::panicking::update_hook::<alloc[d1c3a5de2ae3927d]::boxed::Box<rustc_driver_impl[d0a4257edbfecd8c]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7b6b0a638e76 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hfe27c2dc23f9f057
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/alloc/src/boxed.rs:2036:9
  13:     0x7b6b0a638e76 - std::panicking::rust_panic_with_hook::h6ef218c13552449b
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/std/src/panicking.rs:799:13
  14:     0x7b6b06e70374 - std[aaff33ccdb7ea7e2]::panicking::begin_panic::<rustc_errors[6ff9940fea39db99]::ExplicitBug>::{closure#0}
  15:     0x7b6b06e6cff6 - std[aaff33ccdb7ea7e2]::sys_common::backtrace::__rust_end_short_backtrace::<std[aaff33ccdb7ea7e2]::panicking::begin_panic<rustc_errors[6ff9940fea39db99]::ExplicitBug>::{closure#0}, !>
  16:     0x7b6b06e6ccd6 - std[aaff33ccdb7ea7e2]::panicking::begin_panic::<rustc_errors[6ff9940fea39db99]::ExplicitBug>
  17:     0x7b6b06e79181 - <rustc_errors[6ff9940fea39db99]::diagnostic::BugAbort as rustc_errors[6ff9940fea39db99]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  18:     0x7b6b076ede48 - <rustc_errors[6ff9940fea39db99]::DiagCtxt>::span_bug::<rustc_span[6a84a2dd20c85103]::span_encoding::Span, alloc[d1c3a5de2ae3927d]::string::String>
  19:     0x7b6b0770a73d - rustc_middle[783a1d49f02736b7]::util::bug::opt_span_bug_fmt::<rustc_span[6a84a2dd20c85103]::span_encoding::Span>::{closure#0}
  20:     0x7b6b0770a76a - rustc_middle[783a1d49f02736b7]::ty::context::tls::with_opt::<rustc_middle[783a1d49f02736b7]::util::bug::opt_span_bug_fmt<rustc_span[6a84a2dd20c85103]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  21:     0x7b6b07704c8b - rustc_middle[783a1d49f02736b7]::ty::context::tls::with_context_opt::<rustc_middle[783a1d49f02736b7]::ty::context::tls::with_opt<rustc_middle[783a1d49f02736b7]::util::bug::opt_span_bug_fmt<rustc_span[6a84a2dd20c85103]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  22:     0x7b6b05906ec7 - rustc_middle[783a1d49f02736b7]::util::bug::span_bug_fmt::<rustc_span[6a84a2dd20c85103]::span_encoding::Span>
  23:     0x7b6b0923d880 - <rustc_resolve[956dd1017de6f360]::Resolver>::resolve_crate::{closure#0}
  24:     0x7b6b09234c9c - <rustc_resolve[956dd1017de6f360]::Resolver>::resolve_crate
  25:     0x7b6b0881032b - rustc_interface[3360209938edb6a7]::passes::resolver_for_lowering_raw
  26:     0x7b6b0880f567 - rustc_query_impl[77a161a94332e7e1]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[77a161a94332e7e1]::query_impl::resolver_for_lowering_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[783a1d49f02736b7]::query::erase::Erased<[u8; 16usize]>>
  27:     0x7b6b0880f555 - <rustc_query_impl[77a161a94332e7e1]::query_impl::resolver_for_lowering_raw::dynamic_query::{closure#2} as core[8db7010e98043ac]::ops::function::FnOnce<(rustc_middle[783a1d49f02736b7]::ty::context::TyCtxt, ())>>::call_once
  28:     0x7b6b09002e85 - rustc_query_system[61cd216fcfa07925]::query::plumbing::try_execute_query::<rustc_query_impl[77a161a94332e7e1]::DynamicConfig<rustc_query_system[61cd216fcfa07925]::query::caches::SingleCache<rustc_middle[783a1d49f02736b7]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[77a161a94332e7e1]::plumbing::QueryCtxt, false>
  29:     0x7b6b09002a01 - rustc_query_impl[77a161a94332e7e1]::query_impl::resolver_for_lowering_raw::get_query_non_incr::__rust_end_short_backtrace
  30:     0x7b6b08ea30be - rustc_interface[3360209938edb6a7]::interface::run_compiler::<core[8db7010e98043ac]::result::Result<(), rustc_span[6a84a2dd20c85103]::ErrorGuaranteed>, rustc_driver_impl[d0a4257edbfecd8c]::run_compiler::{closure#0}>::{closure#1}
  31:     0x7b6b08e727c9 - std[aaff33ccdb7ea7e2]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[3360209938edb6a7]::util::run_in_thread_with_globals<rustc_interface[3360209938edb6a7]::util::run_in_thread_pool_with_globals<rustc_interface[3360209938edb6a7]::interface::run_compiler<core[8db7010e98043ac]::result::Result<(), rustc_span[6a84a2dd20c85103]::ErrorGuaranteed>, rustc_driver_impl[d0a4257edbfecd8c]::run_compiler::{closure#0}>::{closure#1}, core[8db7010e98043ac]::result::Result<(), rustc_span[6a84a2dd20c85103]::ErrorGuaranteed>>::{closure#0}, core[8db7010e98043ac]::result::Result<(), rustc_span[6a84a2dd20c85103]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[8db7010e98043ac]::result::Result<(), rustc_span[6a84a2dd20c85103]::ErrorGuaranteed>>
  32:     0x7b6b08e72576 - <<std[aaff33ccdb7ea7e2]::thread::Builder>::spawn_unchecked_<rustc_interface[3360209938edb6a7]::util::run_in_thread_with_globals<rustc_interface[3360209938edb6a7]::util::run_in_thread_pool_with_globals<rustc_interface[3360209938edb6a7]::interface::run_compiler<core[8db7010e98043ac]::result::Result<(), rustc_span[6a84a2dd20c85103]::ErrorGuaranteed>, rustc_driver_impl[d0a4257edbfecd8c]::run_compiler::{closure#0}>::{closure#1}, core[8db7010e98043ac]::result::Result<(), rustc_span[6a84a2dd20c85103]::ErrorGuaranteed>>::{closure#0}, core[8db7010e98043ac]::result::Result<(), rustc_span[6a84a2dd20c85103]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[8db7010e98043ac]::result::Result<(), rustc_span[6a84a2dd20c85103]::ErrorGuaranteed>>::{closure#2} as core[8db7010e98043ac]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  33:     0x7b6b0a642cab - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h1c1c6804813d5e77
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/alloc/src/boxed.rs:2022:9
  34:     0x7b6b0a642cab - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h3d84aa3b40a67b8f
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/alloc/src/boxed.rs:2022:9
  35:     0x7b6b0a642cab - std::sys::pal::unix::thread::Thread::new::thread_start::h60cefba2ee6cce64
                               at /rustc/91d5e4af8674bde06096bac1e7ac04af0e94c691/library/std/src/sys/pal/unix/thread.rs:108:17
  36:     0x7b6b0a3e155a - <unknown>
  37:     0x7b6b0a45ea3c - <unknown>
  38:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.80.0-nightly (91d5e4af8 2024-04-28) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z unstable-options -Z dump-mir-dir=dir

query stack during panic:
#0 [resolver_for_lowering_raw] getting the resolver for lowering
end of query stack
error: aborting due to 1 previous error


```

</p>
</details>

### LINKED ISSUE #125013: ICE:`assertion failed: import.imported_module.get().is_none()`
This crashes with every edition except 2015 :thinking: 

snippet:
````rust
use io::{self as std};
use std::ops::Deref::{self as io};


````

Version information
````
rustc 1.80.0-nightly (35c5e67c6 2024-05-11)
binary: rustc
commit-hash: 35c5e67c69cbde49b47fe537e296803b6a25b456
commit-date: 2024-05-11
host: x86_64-unknown-linux-gnu
release: 1.80.0-nightly
LLVM version: 18.1.4

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Zunstable-options --edition=2024`

<details><summary><strong>Program output</strong></summary>
<p>

```
thread 'rustc' panicked at compiler/rustc_resolve/src/imports.rs:926:21:
assertion failed: import.imported_module.get().is_none()
stack backtrace:
   0:     0x7d29815dfc45 - std::backtrace_rs::backtrace::libunwind::trace::hba6094fc6b596e2e
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/std/src/../../backtrace/src/backtrace/libunwind.rs:105:5
   1:     0x7d29815dfc45 - std::backtrace_rs::backtrace::trace_unsynchronized::h3b202ba307e40438
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7d29815dfc45 - std::sys_common::backtrace::_print_fmt::h86408adb99a13f5c
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x7d29815dfc45 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hbd6f80f0e5997f54
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7d298162ee0b - core::fmt::rt::Argument::fmt::hc498e90c1989f8fc
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/core/src/fmt/rt.rs:165:63
   5:     0x7d298162ee0b - core::fmt::write::ha011919bce0aa4f2
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/core/src/fmt/mod.rs:1157:21
   6:     0x7d29815d4a0f - std::io::Write::write_fmt::hc5d7335558d7c2f5
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/std/src/io/mod.rs:1835:15
   7:     0x7d29815dfa1e - std::sys_common::backtrace::_print::hb5e8494623fbcf06
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7d29815dfa1e - std::sys_common::backtrace::print::h944fe8485fd95854
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7d29815e2389 - std::panicking::default_hook::{{closure}}::hd7a32cc14f6b3aad
  10:     0x7d29815e20cd - std::panicking::default_hook::hce26d1b7cfeb1dbb
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/std/src/panicking.rs:298:9
  11:     0x7d297ded14af - std[51ea32a909324831]::panicking::update_hook::<alloc[77c3f3c3161b06c0]::boxed::Box<rustc_driver_impl[aad4619bae6cbfbd]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7d29815e2a86 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h3bc218da95aac6f0
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/alloc/src/boxed.rs:2036:9
  13:     0x7d29815e2a86 - std::panicking::rust_panic_with_hook::h104a5f4852f71fae
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/std/src/panicking.rs:799:13
  14:     0x7d29815e27fb - std::panicking::begin_panic_handler::{{closure}}::hbe2cffd5894aac21
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/std/src/panicking.rs:656:13
  15:     0x7d29815e0109 - std::sys_common::backtrace::__rust_end_short_backtrace::h124d376bffe44b06
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/std/src/sys_common/backtrace.rs:171:18
  16:     0x7d29815e2567 - rust_begin_unwind
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/std/src/panicking.rs:652:5
  17:     0x7d298162b3d3 - core::panicking::panic_fmt::ha6a495494fa26de2
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/core/src/panicking.rs:72:14
  18:     0x7d298162b47c - core::panicking::panic::h43564003bddcc9c0
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/core/src/panicking.rs:146:5
  19:     0x7d2980257274 - <rustc_resolve[287ffcdb65d0abd4]::Resolver>::resolve_crate::{closure#0}
  20:     0x7d2980248ac0 - <rustc_resolve[287ffcdb65d0abd4]::Resolver>::resolve_crate
  21:     0x7d297f7be7c1 - rustc_interface[1af8fbac95f387b0]::passes::resolver_for_lowering_raw
  22:     0x7d297f7bda0d - rustc_query_impl[b785da3d31912d7f]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[b785da3d31912d7f]::query_impl::resolver_for_lowering_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1afb1a57a12263fe]::query::erase::Erased<[u8; 16usize]>>
  23:     0x7d297f7bd9ef - <rustc_query_impl[b785da3d31912d7f]::query_impl::resolver_for_lowering_raw::dynamic_query::{closure#2} as core[7b5ea08294f46cd8]::ops::function::FnOnce<(rustc_middle[1afb1a57a12263fe]::ty::context::TyCtxt, ())>>::call_once
  24:     0x7d297ffa34c5 - rustc_query_system[42f095e9d6ad5ed8]::query::plumbing::try_execute_query::<rustc_query_impl[b785da3d31912d7f]::DynamicConfig<rustc_query_system[42f095e9d6ad5ed8]::query::caches::SingleCache<rustc_middle[1afb1a57a12263fe]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[b785da3d31912d7f]::plumbing::QueryCtxt, false>
  25:     0x7d297ffa3059 - rustc_query_impl[b785da3d31912d7f]::query_impl::resolver_for_lowering_raw::get_query_non_incr::__rust_end_short_backtrace
  26:     0x7d297fe3bffe - rustc_interface[1af8fbac95f387b0]::interface::run_compiler::<core[7b5ea08294f46cd8]::result::Result<(), rustc_span[f63607c00cb4b4d8]::ErrorGuaranteed>, rustc_driver_impl[aad4619bae6cbfbd]::run_compiler::{closure#0}>::{closure#1}
  27:     0x7d297fe27d09 - std[51ea32a909324831]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[1af8fbac95f387b0]::util::run_in_thread_with_globals<rustc_interface[1af8fbac95f387b0]::util::run_in_thread_pool_with_globals<rustc_interface[1af8fbac95f387b0]::interface::run_compiler<core[7b5ea08294f46cd8]::result::Result<(), rustc_span[f63607c00cb4b4d8]::ErrorGuaranteed>, rustc_driver_impl[aad4619bae6cbfbd]::run_compiler::{closure#0}>::{closure#1}, core[7b5ea08294f46cd8]::result::Result<(), rustc_span[f63607c00cb4b4d8]::ErrorGuaranteed>>::{closure#0}, core[7b5ea08294f46cd8]::result::Result<(), rustc_span[f63607c00cb4b4d8]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[7b5ea08294f46cd8]::result::Result<(), rustc_span[f63607c00cb4b4d8]::ErrorGuaranteed>>
  28:     0x7d297fe27ab6 - <<std[51ea32a909324831]::thread::Builder>::spawn_unchecked_<rustc_interface[1af8fbac95f387b0]::util::run_in_thread_with_globals<rustc_interface[1af8fbac95f387b0]::util::run_in_thread_pool_with_globals<rustc_interface[1af8fbac95f387b0]::interface::run_compiler<core[7b5ea08294f46cd8]::result::Result<(), rustc_span[f63607c00cb4b4d8]::ErrorGuaranteed>, rustc_driver_impl[aad4619bae6cbfbd]::run_compiler::{closure#0}>::{closure#1}, core[7b5ea08294f46cd8]::result::Result<(), rustc_span[f63607c00cb4b4d8]::ErrorGuaranteed>>::{closure#0}, core[7b5ea08294f46cd8]::result::Result<(), rustc_span[f63607c00cb4b4d8]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[7b5ea08294f46cd8]::result::Result<(), rustc_span[f63607c00cb4b4d8]::ErrorGuaranteed>>::{closure#2} as core[7b5ea08294f46cd8]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  29:     0x7d29815ec8cb - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h47404fef54470e82
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/alloc/src/boxed.rs:2022:9
  30:     0x7d29815ec8cb - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h94736f8ab20acaae
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/alloc/src/boxed.rs:2022:9
  31:     0x7d29815ec8cb - std::sys::pal::unix::thread::Thread::new::thread_start::h3dd6ac2255354521
                               at /rustc/35c5e67c69cbde49b47fe537e296803b6a25b456/library/std/src/sys/pal/unix/thread.rs:108:17
  32:     0x7d298138b55a - <unknown>
  33:     0x7d2981408a3c - <unknown>
  34:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.80.0-nightly (35c5e67c6 2024-05-11) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z unstable-options -Z dump-mir-dir=dir

query stack during panic:
#0 [resolver_for_lowering_raw] getting the resolver for lowering
end of query stack

```

</p>
</details>

### PR DESCRIPTION
**Pull request by bvanjoi**
- Fixes #124490
- Fixes #125013

This issue arises from incorrect resolution updates, for example:

```rust
mod a {
    pub mod b {
        pub mod c {}
    }
}

use a::*;

use b::c;
use c as b;

fn main() {}
```

1. In the first loop, binding `(root, b)` is refer to `root::a::b` due to `use a::*`. 
    1. However, binding `(root, c)` isn't defined by `use b::c` during this stage because `use c as b` falls under the `single_imports` of `(root, b)`, where the `imported_module` hasn't been computed yet. This results in marking the `path_res` for `b` as `Indeterminate`.
    2. Then, the `imported_module` for `use c as b` will be recorded.
2. In the second loop, `use b::c` will be processed again:
    1. Firstly, it attempts to find the `path_res` for `(root, b)`.
    2. It will iterate through the `single_imports` of `use b::c`, encounter `use c as b`, attempt to resolve `c` in `root`, and ultimately return `Err(Undetermined)`, thus passing the iterator.
    3. Use the binding `(root, b)` -> `root::a::b` introduced by `use a::*` and ultimately return `root::a::b` as the `path_res` of `b`.
    4. Then define the binding `(root, c)` -> `root::a::b::c`.
3. Then process `use c as b`, update the resolution for `(root, b)` to refer to `root::a::b::c`, ultimately causing inconsistency.

In my view, step `2.2` has an issue where it should exit early, similar to the behavior when there's no `imported_module`. Therefore, I've added an attribute called `indeterminate` to `ImportData`. This will help us handle only those single imports that have at least one determined binding.

r? @petrochenkov

### CODE REVIEW

**Comment by bvanjoi** [comment]
> https://github.com/rust-lang/rust/pull/124840#issuecomment-2098053650

Can't reproduce this locally. Anyway, mark it as a draft for now.

**Comment by bvanjoi** [comment]
Update: I've changed that it now only returns ﻿`Undetermined` when resolving modules.

I'm not sure if there's a better solution or potential edge cases. Regardless, since CI is passing, mark it as ready for review.

**Comment by bvanjoi** [comment]
> Does resolve_ident_in_module(...) returns Err(Determined)

Yep, it will return ﻿`Err(Determined)` for `﻿resolve_ident_in_module(c, root)`. It might be a typo..

**Comment by bvanjoi** [comment]
The reason why ﻿`resolve_ident_in_module(c, root)` returns ﻿`Err(Determined)` is because it skips the iteration of ﻿`single_imports`. More specifically, the `﻿single_import.vis` value is `﻿None` because it was already taken at the [entry point](https://github.com/rust-lang/rust/blob/master/compiler/rustc_resolve/src/imports.rs#L774). This results in hitting a ﻿`continue` statement.

**Comment by bvanjoi** [comment]
Also, I tried moving the ﻿`single_import.imported_module.get()` condition before ﻿`single_import.vis.get()`, but `﻿./x.py` build failed to compile. Therefore, I've decided not to solve this problem in this manner. :E

**Comment by petrochenkov** [comment]
I suspect that resolver may be stuck on some valid code after this change, let's check.
@bors try

**Comment by petrochenkov** [comment]
@craterbot check

File: compiler/rustc_resolve/src/ident.rs [hunk 1]
```diff
@@ -979,15 +994,22 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
             let Some(module) = single_import.imported_module.get() else {
                 return Err((Undetermined, Weak::No));
             };
-            let ImportKind::Single { source: ident, .. } = single_import.kind else {
+            let ImportKind::Single { source: ident, source_bindings, .. } = &single_import.kind
+            else {
                 unreachable!();
             };
+            if binding.map_or(false, |binding| binding.module().is_some())
+                && source_bindings.iter().all(|binding| matches!(binding.get(), Err(Undetermined)))
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/ident.rs:1006]: Could you add a comment here explaining why this is done?

**Comment by petrochenkov** [comment]
r=me with the comment https://github.com/rust-lang/rust/pull/124840#discussion_r1624639116.
@rustbot author

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Comment by fmease** [comment]
bors sleepy @bors r-

**Pull request merged**